### ISSUE TITLE
Warn on multiple instances of StyleSheetRegistry

### ISSUE THREAD

**Issue by jamestalmage** [zeit/styled-jsx#533]
Due to a SemVer mismatch between `Next` and my own `package.json`, I ended up with the following directory structure:

```
node_modules/
├── next/
│   └── node_modules/
│       ├── styled-jsx/
│       └── .....
└── styled-jsx/
```

This led to `Next` not rendering my styles server side (it's registry was empty).

I see two options: 
  1. Set a global flag and warn if two different registries are instantiated.
  2. Drop the built-in support from `Next`, and provide an easy solution for injecting it (similar to `withPreact`, etc).

I guess you could also use a singleton registry, but that's problematic.

**Comment by jamestalmage** [comment]
Option 1 is pretty straightforward. I'd happily submit a PR.

**Comment by jamestalmage** [comment]
3rd Option:

Build detection of this problem into the `Next` server.

**Comment by giuseppeg** [comment]
@jamestalmage thanks for the detailed description. We have an open issue to fix this https://github.com/zeit/styled-jsx/issues/64 but it requires some work and I am not sure when we will get to it. Probably now that we have the new context API it is easier to implement that, let me know if you'd be interested in contributing with a solution.

**Comment by jamestalmage** [comment]
I would happily contribute. Given #64, it seems that a singleton registry isn't going to be ideal in all scenarios. 

I am guessing that we can safely assume the user doesn't want to be use a deep copy of `styled-jsx` if another version exists nearer the root of the node_modules heirarchy. 

I propose we walk the heirarchy and simply issue a warning on the console if we find evidence that user code might be accessing a different version.

This check should probably be shimmed with a no-op for the browser, as it involves the file system

**Comment by giuseppeg** [comment]
> I am not suggesting we actually install the registry globally, but just set a flag and issue a warning.

yep I know but it still feels dirty - nothing wrong with the solutions you suggested, I just think that the only way to "fix" this is to go for an hack and I think it is not super important. I mean imagine if any lib that could be installed multiple times added checks of this kind.

> that would be hard to do reliably. If you depend on something that depends on styled-jsx

other libraries should declare `styled-jsx` as a `peerDependencies`, when they don't do this and declare it as a regular dependency they are doing something wrong in my opinion. Next.js is doing this but that's their project and they want to make it easier to use styled-jsx. At least though they could add the check I was talking about.

> I don't think the babel plugin check is necessarily the way to go either

I am suggesting that Next.js checks the babelrc configuration to see if styled-jsx is included twice

**Comment by giuseppeg** [comment]
James I hope my comments didn’t sound too harsh. If they did sorry for the tone :) I was typing from my phone and didn’t read again before publishing. I just wanted to explain the rationale that makes me say “in my opinion we shouldn’t implement this”. If you feel strong about this feature maybe we can open an issue on the next.js repo and see what the zeit folks say

**Comment by timneutkens** [comment]
Seems like none of the mentioned solutions are considering the current flushing mechanism has to be changed anyway to be async-safe anyway.

Basically it would make a lot of sense to have styled-jsx emit style tags while rendering through a function passed into context. Similar to what react-loadable does:

https://github.com/zeit/next.js/blob/canary/packages/next-server/server/render.tsx#L172

This way there's a buffer of tags per request, instead of in the global module.

**Comment by giuseppeg** [comment]
@timneutkens isn't that what is happening right now? https://github.com/zeit/styled-jsx/blob/master/src/style.js the problem is that styleSheetRegistry is not injectable via context, if we do so we should be good

**Comment by timneutkens** [comment]
Exactly, when the registry can come from context that'd be the solution 👍

**Comment by giuseppeg** [comment]
Cool we will fix this with #64 then 🎉 thank you folks for your inputs
